### PR TITLE
feat: add helpful error messages for removed plugins

### PR DESCRIPTION
**Pull request by kingdaro**
**What kind of change does this PR introduce?**

Gives helpful error messages when accessing removed plugins from v3.

TODO:
- [x] Tests
- [x] ~~Include links to relevant documentation~~ Will create an issue on webpack.js.org repo to track this

**Did you add tests for your changes?**

Yes

**Summary**

The current behavior will produce an error like "TypeError: webpack.optimize.UglifyJsPlugin is not a constructor". This PR gives a more appropriate message to point the user in the right direction. 

See https://github.com/webpack/webpack/issues/6409#issuecomment-361338563

**Does this PR introduce a breaking change?**

No

### CODE REVIEW

File: lib/webpack.js [hunk 1]
```diff
@@ -132,3 +132,23 @@ exportPlugins(exports.node = {}, {
 exportPlugins(exports.debug = {}, {
 	"ProfilingPlugin": () => require("./debug/ProfilingPlugin"),
 });
+
+const defineMissingPluginError = (pluginName, errorMessage) => {
+	Reflect.defineProperty(exports.optimize, pluginName, {
+		configurable: false,
+		enumerable: true,
+		get() {
+			throw new Error(errorMessage);
```
Comment (TheLarkInn) [inline lib/webpack.js]:
```
throw new Error(`RemovedPluginError: ${errorMessage}`);
```

File: lib/webpack.js [hunk 2]
```diff
@@ -132,3 +132,23 @@ exportPlugins(exports.node = {}, {
...
+		}
+	});
+};
+
+defineMissingPluginError(
```
Comment (ooflorent) [inline lib/webpack.js]: Could you add the following comment before each call?

```js
// TODO remove in webpack 5
```

File: lib/webpack.js [hunk 3]
```diff
@@ -132,3 +132,23 @@ exportPlugins(exports.node = {}, {
 exportPlugins(exports.debug = {}, {
 	"ProfilingPlugin": () => require("./debug/ProfilingPlugin"),
 });
+
+const defineMissingPluginError = (pluginName, errorMessage) => {
+	Reflect.defineProperty(exports.optimize, pluginName, {
```
Comment (ooflorent) [inline lib/webpack.js]: Not available in node 6. Could you use `Object.defineProperty` instead?

Comment (kingdaro) [reply lib/webpack.js]: Reflect actually does appear to be in node 6. I used NVM to test:
```
PS C:\Users\kingd> nvm install 6.0.0
Downloading node.js version 6.0.0 (64-bit)...
Complete
Creating C:\ProgramData\nvm\temp

Downloading npm version 3.8.6... Complete
Installing npm v3.8.6...

Installation complete. If you want to use this version, type

nvm use 6.0.0
PS C:\Users\kingd> nvm use 6.0.0
Now using node v6.0.0 (64-bit)
PS C:\Users\kingd> node -v
v6.0.0
PS C:\Users\kingd> node
> Reflect
Reflect
{}
> Reflect.defineProperty
[Function: defineProperty]
```

Comment (kingdaro) [reply lib/webpack.js]: Reflect is preferred since it doesn't throw like `Object.defineProperty` does.

File: lib/webpack.js [hunk 4]
```diff
@@ -132,3 +132,25 @@ exportPlugins(exports.node = {}, {
...
+		configurable: false,
+		enumerable: true,
+		get() {
+			throw new Error(`RemovedPluginError: ${errorMessage}`);
+		}
+	});
+};
+
+// TODO remove in webpack 5
+defineMissingPluginError(
```
Comment (sokra) [inline lib/webpack.js]: pass `exports.optimize` as first argument, so we could define missing plugins for other categories too.

File: lib/webpack.js [hunk 5]
```diff
@@ -132,3 +132,25 @@ exportPlugins(exports.node = {}, {
 exportPlugins(exports.debug = {}, {
 	"ProfilingPlugin": () => require("./debug/ProfilingPlugin"),
 });
+
+const defineMissingPluginError = (pluginName, errorMessage) => {
+	Reflect.defineProperty(exports.optimize, pluginName, {
+		configurable: false,
+		enumerable: true,
+		get() {
+			throw new Error(`RemovedPluginError: ${errorMessage}`);
```
Comment (sokra) [inline lib/webpack.js]: Make it a Error class: RemovedPluginError

Comment (kingdaro) [reply lib/webpack.js]: Should I write that inline here, or put it in a separate file somewhere?

File: lib/webpack.js (hunk 3 above)
Comment (ooflorent) [reply lib/webpack.js]: Nevermind. When zoomed in, node.green table is messed up.

**Comment by ooflorent** [comment]
You can't test this using an integration test. You should use a unit test.

**Comment by kingdaro** [comment]
Thanks, I did mean for it to be a unit test, but I initially misunderstood where it was supposed to go.

**Comment by ooflorent** [comment]
> I initially misunderstood where it was supposed to go.

webpack test suite is not really straightforward to work with 😅

**Comment by sokra** [comment]
Thanks

**Comment by TheLarkInn** [comment]
Congrats on your first webpack contribution!!!!!! 🎉🎉🎉🎉🎉

**Pull request merged**